### ISSUE TITLE
index in case of (nested) csv

### ISSUE THREAD

**Issue by elsdvlee** [kg-construct/rml-lv#20]
How to calculate the index in case of csv? 
Until now I supposed that we keep the index of the row
| # | name.# | name   | birthyear.# | birthyear |
| - | ------ | ------ | ----------- | --------- |
| 0 | 0      | alice  | 0           | 1995      |
| 1 | 1      | bob    | 1           | 1999      |
| 2 | 2      | tobias | 2           | 2005      |

However, is this in line with how we calculated the index for json file? There the index is 0 if there is only on value for the logical iteration. In case of csv there standard only one value for the logical iteration, so then the index of all fields should be 0, except for the overall index #. 

| # | name.# | name   | birthyear.# | birthyear |
| - | ------ | ------ | ----------- | --------- |
| 0 | 0      | alice  | 0           | 1995      |
| 1 | 0    | bob    | 0           | 1999      |
| 2 | 0      | tobias | 0          | 2005      |

And what to do with a nested csv column? In that case it makes more sense to maintain the index of the row, otherwise we cannot reconstruct the order. 
```
{
  "people": [
    {
      "name": "alice",
      "hobbies": "id;type\n1;volleybal\n2;basketball\n3;horses"
    },
    {
      "name": "bob",
      "hobbies": "id;type\n1;football"
    }
  ]
}
```
| # | name  | name.# | hobbies | hobbies.# | hobbies.id | hobbies.id.# | hobbies.type | hobbies.type.# |
| - | ----- | ------ | ------- | --------- | ---------- | ------------ | ------------ | -------------- |
| 0 | alice | 0      | (...)   | 0         | 1          | 0            | volleybal    | 0              |
| 0 | alice | 0      | (...)   | 0         | 2          | 1            | basketball   | 1              |
| 0 | alice | 0      | (...)   | 0         | 3          | 2            | horses       | 2              |
| 1 | bob   | 0      | (...)   | 0         | 1          | 0            | football     | 0              |

Should we define the algorithm to calculate the index for a field per reference formulation?

**Comment by elsdvlee** [comment]
>>otherwise we cannot reconstruct the order.

>@elsdvlee what did you mean by this?
I mean the support for collections and containers, reconstruct a list in the original order.

**Comment by elsdvlee** [comment]
So, if I understand you correctly, in case of csv the correct indexes are: 
| # | name.# | name   | birthyear.# | birthyear |
| - | ------ | ------ | ----------- | --------- |
| 0 | 0      | alice  | 0           | 1995      |
| 1 | 1      | bob    | 1           | 1999      |
| 2 | 2      | tobias | 2           | 2005      |

In case of json, you restart counting from 0 as a result of the json path expression used as reference. 
| # | name  | name.# | hobbies | hobbies.# | hobbies.id | hobbies.id.# | hobbies.type | hobbies.type.# |
| - | ----- | ------ | ------- | --------- | ---------- | ------------ | ------------ | -------------- |
| 0 | alice | 0      | (...)   | 0         | 1          | 0            | volleybal    | 0              |
| 0 | alice | 0      | (...)   | 0         | 2          | 1            | basketball   | 1              |
| 0 | alice | 0      | (...)   | 0         | 3          | 2            | horses       | 2              |
| 1 | bob   | 0      | (...)   | 0         | 1          | 0            | football     | 0              |

**Comment by bjdmeest** [comment]
No, for CSV the correct indexes are, I think:

| # | name.# | name   | birthyear.# | birthyear |
| - | ------ | ------ | ----------- | --------- |
| 0 | 0      | alice  | 0           | 1995      |
| 1 | 0      | bob    | 0           | 1999      |
| 2 | 0      | tobias | 0           | 2005      |

Because it's each time the 'first' name/birthyear of that row

**Comment by elsdvlee** [comment]
That is exactly my point. In case of csv inside json (mixed format) we skip one step: we make additional iterations and from those iterations we take the reference, so should we then have 0 as index here as well? In that case we cannot know the order anymore. 
| # | name  | name.# | hobbies | hobbies.# | hobbies.id | hobbies.id.# | hobbies.type | hobbies.type.# |
| - | ----- | ------ | ------- | --------- | ---------- | ------------ | ------------ | -------------- |
| 0 | alice | 0      | (...)   | 0         | 1          | 0            | volleybal    | 0              |
| 0 | alice | 0      | (...)   | 0         | 2          | 0            | basketball   | 0            |
| 0 | alice | 0      | (...)   | 0         | 3          | 0            | horses       | 0              |
| 1 | bob   | 0      | (...)   | 0         | 1          | 0            | football     | 0              |

**Comment by bjdmeest** [comment]
I would expect that hobbies.# then _would_ increment, because `hobbies` doesn't result in a single field, but in multiple iterations

**Comment by pmaria** [comment]
@elsdvlee It is hard to follow your example without the field definitions.

In general, this is how it should work:

A field's index is coupled to its parent. When you create a new child field for a "parent" the index for the child field starts at 0, and will increment for each item in the expression result value of that child field's expression.

**Comment by elsdvlee** [comment]
@pmaria here is the corresponding rml mapping: 
```
@prefix rml: <http://w3id.org/rml/> .
@prefix : <http://example.org/> .

:jsonSource a rml:LogicalSource ;
  rml:source "./test_cases/json_csv_data.json";
  rml:referenceFormulation rml:JSONPath ;
  rml:iterator "$.people[*]" .

:jsonView a rml:LogicalView ;
  rml:onLogicalSource :jsonSource ;
  rml:field [
    rml:fieldName "name" ;
    rml:reference "$.name" ;
  ] ;
  rml:field [
    rml:fieldName "hobbies" ;
    rml:reference "$.hobbies" ;
    rml:referenceFormulation rml:CSV;
    rml:field [
      rml:fieldName "type" ;
      rml:reference "type" ;
    ] ;
    rml:field [
      rml:fieldName "id" ;
      rml:reference "id" ;
    ] ;
  ] .
```

**Comment by pmaria** [comment]
Thanks @elsdvlee, I now see your point. 

When we switch reference formulation the iteration context also changes.
I think we need to explore this a bit further, but I can imagine that when switching reference formulation we could also require(?) specifying iterator of use the default for the reference formulation.

So then you'd get:

```turtle
@prefix rml: <http://w3id.org/rml/> .
@prefix : <http://example.org/> .

:jsonSource a rml:LogicalSource ;
  rml:source "./test_cases/json_csv_data.json";
  rml:referenceFormulation rml:JSONPath ;
  rml:iterator "$.people[*]" .

:jsonView a rml:LogicalView ;
  rml:onLogicalSource :jsonSource ;
  rml:field [
    rml:fieldName "name" ;
    rml:reference "$.name" ;
  ] ;
  rml:field [
    rml:fieldName "hobbies" ;
    rml:reference "$.hobbies" ;
    rml:referenceFormulation rml:CSV;

    # implicit iterator is CSV row
    
    rml:field [
      rml:fieldName "type" ;
      rml:reference "type" ;
    ] ;
    rml:field [
      rml:fieldName "id" ;
      rml:reference "id" ;
    ] ;
  ] .
```

```json
{
  "people": [
    {
      "name": "alice",
      "hobbies": "id;type\n1;volleybal\n2;basketball\n3;horses"
    },
    {
      "name": "bob",
      "hobbies": "id;type\n1;football"
    }
  ]
}
```

| # | name  | name.# | hobbies | hobbies.# | hobbies.id | hobbies.id.# | hobbies.type | hobbies.type.# |
| - | ----- | ------ | ------- | --------- | ---------- | ------------ | ------------ | -------------- |
| 0 | alice | 0      | (...)   | 0         | 1          | 0            | volleybal    | 0              |
| 0 | alice | 0      | (...)   | 1         | 2          | 0            | basketball   | 0              |
| 0 | alice | 0      | (...)   | 2         | 3          | 0            | horses       | 0              |
| 1 | bob   | 0      | (...)   | 0         | 1          | 0            | football     | 0              |

**Comment by bjdmeest** [comment]
Agree with the latest comment of Pano: include iterator every time you change reference formulations (or use the default if there is one).

**Comment by pmaria** [comment]
agreed in CG meeting 03/20/2024 that this approach is a good solution
